Packages/com.unity.inputsystem/InputSystem/Devices/ScreenKeyboard.cs

```diff
@@ -0,0 +1,290 @@
+    {
+        protected ScreenKeyboardState m_KeyboardState;
+        protected ScreenKeyboardShowParams m_ShowParams;
+        private ScreenKeyboardCallbacks m_Callbacks;
+
+        /// <summary>
+        /// Returns the state of the keyboard
+        /// </summary>
+        public ScreenKeyboardState state
+        {
+            get => m_KeyboardState;
+        }
+
+        /// <summary>
+        /// Cleanup any platform specific resources
+        /// </summary>
+        public virtual void Dispose()
+        {
+        }
+
+        /// <summary>
+        /// Show the screen keyboard with customized options.
+        /// </summary>
+        /// <param name="showParams"></param>
+        /// <param name="callbacks"></param>
+        public void Show(ScreenKeyboardShowParams showParams = default, ScreenKeyboardCallbacks callbacks = default)
+        {
+            m_ShowParams = showParams;
+            m_Callbacks = callbacks;
+            if (m_ShowParams.initialText == null)
+                m_ShowParams.initialText = String.Empty;
+            if (m_ShowParams.placeholderText == null)
+                m_ShowParams.placeholderText = String.Empty;
+            InternalShow();
+        }
+
+        /// <summary>
+        /// Hide screen keyboard
+        /// </summary>
+        public void Hide()
+        {
+            InternalHide();
+        }
+
+        protected abstract void InternalShow();
+
+        protected abstract void InternalHide();
+
+        protected void ReportInputFieldChange(string text)
+        {
+            m_Callbacks.inputFieldTextChanged?.Invoke(text);
+        }
+
+        protected void ReportStateChange(ScreenKeyboardState state)
+        {
+            if (state != m_KeyboardState)
+            {
+                m_KeyboardState = state;
+                m_Callbacks.stateChanged?.Invoke(state);
+            }
+        }
+
+        protected void ReportSelectionChange(int start, int length)
+        {
+            if (m_KeyboardState != ScreenKeyboardState.Visible)
+                return;
+
+            var selection = new RangeInt(start, length);
+            m_Callbacks.inputFieldSelectionChanged?.Invoke(selection);
+        }
+
+        /// <summary>
+        /// Modifies text in screen keyboard's input field.
+        /// You can use this propery even if input field is hidden
+        /// </summary>
+        public abstract string inputFieldText { set; get; }
+
+        /// <summary>
+        /// Returns portion of the screen which is covered by the keyboard.
+        /// </summary>
+        public virtual Rect occludingArea => Rect.zero;
+
+        /// <summary>
+        /// Modify selection of an input field text.
+        /// When input field is hidden, this property does nothing and always returns (text.length, 0)
+        /// If you specify a selection out of bounds of text length, the selection will not be set.
+        /// </summary>
+        public abstract RangeInt selection { set; get; }
```

Comment:
While trying to fix a bug in our existing TouchScreenKeyboard I encountered one quite fundamental issue - we should probably consider having ability to atomically change multiple properties.
For example, when one writes code like this:
keyboard.inputFieldText = newText;
keyboard.selection = new RangeInt(0, 5);

one would expect the keyboards text and selection to be set. However, with keyboard running on different thread than this code, user can press a button on a keyboard exactly in between these two calls leading to funny outcome.